### PR TITLE
Day3: Flatten Binary Tree to Linked List

### PR DESCRIPTION
**Pull request by alberthu16**
Please leave comments, suggestions, tips n tricks!

### CODE REVIEW

File: day3/README.md [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+it would store up to `log(n)` references to nodes. So it's not
+`O(1)` space, I guess.
+
+## Code ##
+
+[Python](./flatten.py)
+
+## Follow Up ## 
+
+(TODO) I'm pretty sure I can use recursion in this problem..
```
Comment (phorust) [inline day3/README.md]: interesting that you went to iterative and have the `store = [ ]` - my initial reaction to the problem statement is that it's super recursive.. but maybe actually writing out some code might've changed my mind

File: day3/flatten.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    def __init__(self, x):
+        self.val = x
+        self.left = None
+        self.right = None
+
+## My Solution ##
+class Solution(object):
+    def flatten(self, root):
+        """
+        :type root: TreeNode
```
Comment (phorust) [inline day3/flatten.py]: what syntax is this? (can I still not make multi line comments?)

File: day3/README.md (hunk 1 above)
Comment (alberthu16) [inline day3/README.md]: My intuition was recursion too.. and then when I tried to write it out I struggled and the iterative approach made more sense to me. I'll definitely revisit and try to do it recursively to practice my recursion chops.

**Comment by phorust** [comment]
curious, do you write test cases before or after you write the solution?

this solution ended up being surprisingly small and elegant! i like it.

File: day3/flatten.py (hunk 2 above)
Comment (alberthu16) [inline day3/flatten.py]: Do you mean for the docstrings? I just copied from the leetcode question: https://leetcode.com/problems/flatten-binary-tree-to-linked-list/#/description

**Comment by alberthu16** [comment]
@phorust thanks! I usually write em before, but it also depends on how lazy I am and how easy the tests are to write in comparison to the code. Like for this problem the tests felt 100x longer to write compared to the actual solution code. Which maybe is generally true LOL

File: day3/README.md [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+           \
+            5
+             \
+              6
+               \
+                7
+
+So the store would be a stack, and I think in the worst case
+it would store up to `log(n)` references to nodes. So it's not
+`O(1)` space, I guess.
```
Comment (alberthu16) [inline day3/README.md]: I'm also confused about the space complexity with using a stack and doing this iteratively. Anyone have tips on how to think about this?

File: day3/flatten.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+            return False
+        tree = tree.right
+        checker += 1
+
+    return True
+
+# TODO(albert): test more edge cases
+
+def tests():
+    print(test_create_preorder_binary_tree())
```
Comment (jjwon0) [inline day3/flatten.py]: you could use asserts instead of prints for even more nice-ness

Comment (jjwon0) [inline day3/flatten.py]: or maybe even have a `test_utils.py` in the root dir

Comment (alberthu16) [inline day3/flatten.py]: gr8 idea

Comment (alberthu16) [inline day3/flatten.py]: That makes me think.. I need to find some way to automate all the boilerplate-y things I need to do for these daily challenges.

- README formatting
- Creating the Day _ directory with a test suite
- Opening a PR
